### ISSUE TITLE
How would you handle recording and playback of bundles from file?

### ISSUE THREAD

**Issue by 0x80** [colinbdclark/osc.js#60]
For an application I would like to record and playback OSC data in javascript. As I understand writing/reading bundles from disk is outside of the scope of this library, but I was hoping someone can give some pointers on how to approach this in JS.

The recording can be done server side (Node.js) and needs to be serialized to file. The payback part reads from that file. It would preferably happen in the browser, but I can stick to server side for this as well.

I could record an array of osc bundles, but it's unclear to me how to convert that to a file format. I imagine I'd use Blob to convert it to binary for file and Filereader to get it back into javascript? But when Blob converts it to binary all the bundles are going to end up on one pile and I don't see how the Filereader API can then tell them apart...

Do you have an idea? Has there been some effort already in this direction with Javascript? I couldn't find any so far.

**Comment by colinbdclark** [comment]
Hi,

Really interesting question. There are a few different approaches you could take. If you simply want to save messages as binary into a file, you'll need to do some kind of framing to be able to distinguish one packet from another. The [OSC best practice](http://lac.linuxaudio.org/2010/papers/37.pdf) for this seems to be to use SLIP to frame messages. I wrote [slip.js](https://github.com/colinbdclark/slip.js) for this purpose, and use it for osc.js' Serial and TCP transports. You could probably write your own <code>osc.Port</code> implementation that extends from <code>osc.SLIPPort</code>, which writes to and reads from files or <code>Blobs</code>. It would be an interesting thing to do, and I'd happily lend a hand or take a pull request if you have something that others can benefit from.

Another approach if you don't need your file format to be directly compatible with other OSC applications (i.e. read by them), and which I think is ultimately more interesting, is to store your OSC messages as JSON and stringify them either into a text file or perhaps into a JSON document store like local data storage, PouchDB, or CouchDB on the server. This is one of the reasons I designed the osc.js architecture in the way that I did--so its messages were very interoperable.

That said, currently not all types of osc.js messages are directly JSON-compatible, since it uses typed arrays for OSC <code>blob</code> and <code>MIDI</code> types. These would need to be converted to standard JS arrays prior to stringifying them. I should really create an osc.js-standard utility for converting messages to JSON-compliant formats. You'll also want to make sure that you're using the <code>metadata: true</code> option so that all the appropriate type information is retained when you stringify your messages.

These are a few ideas off the top of my head. I'd be interested to hear the approach you take, and let me know if you want a hand or have any other questions about it. This sounds like a cool project you're working on!

**Comment by 0x80** [comment]
Hi Colin,

Thanks for the long reply. This slip thing was the missing piece for me. And the OSC best practices is an interesting read.

I'm working on a realtime visual composition tool using browser layout techniques, to be controlled with midi and other data sources. The same data that drives the visuals also creates sound. I want the captured composition (consisting of a sound file + control data) to play back in the browser for "offline" listening.

At the moment my only source is midi so I decided to record / playback that first, and do the mapping on the played back data. For now I'll just timestamp the midi messages and put them in a plain array converted to JSON. I calculated that I can fit about 100k messages into 2mb of gzipped data which should be enough for a start.

Ultimately I'd want the recorded control data to be more descriptive and flexible then plain midi, so I want to use OSC for it. Handling OSC as plain javascript objects seems very memory consuming, that's why I thought about recording them as binary data and unpacking them one by one in realtime, as opposed to parsing a whole recording to/from JSON at once. 

Maybe it's not so bad for transfer bandwidth if you use gzip on the recorded data, as a lot of OSC messages will use the same address strings, but to have all of the data sitting in memory at once doesn't seem very resource efficient to me. Especially since I want to support mobile devices for playback.

It will take me a while to get to this point, but if I have something interesting to share I will let you know :)
